### ISSUE TITLE
Add openCursor(key, primaryKey) and continuePrimaryKey

### ISSUE THREAD

**Issue by inexorabletash** [w3c/IndexedDB#14]
https://www.w3.org/Bugs/Public/show_bug.cgi?id=20257

**Comment by inexorabletash** [comment]
Need to specify what happens if you call `continuePrimaryKey()` on a cursor where the source is an object store. Throw?

**Comment by inexorabletash** [comment]
What should the behavior of `continuePrimaryKey()` be with `"prevunique"` iteration? Given this index:

* r1 = key: 'a', primary key: 1
* r2 = key: 'a', primary key: 2
* r3 = key: 'b', primary key: 1

If direction is "prevunique" and the cursor is at r3, `continue('a')` is specified to end up at r1 (the _first_ record with matching key).

Cases: 
* If direction is "prevunique" and the cursor is at r3, `continue('a', 2)` should end up at... ?
 * r2 seems to make the most sense here
* If direction is "prevunique" and the cursor is at r3, `continue('a', 'a3')` should end up at... ?
 * to me, r2 seems sensible here, but my intuition may be wonky

**Comment by inexorabletash** [comment]
Heh, http://crbug.com/497454

**Comment by inexorabletash** [comment]
Initial work in a4e4d2b945b4abbd1f869b411f37f8ccfe2e0615

**Comment by inexorabletash** [comment]
Another option is to throw if `continuePrimaryKey()` is used with "nextunique" or "prevunique" since the behavior is a bit odd.

**Comment by yathit** [comment]
@inexorabletash +1 throw InvalidAccessError

**Comment by inexorabletash** [comment]
@yathit do you mean for the object store source case, or for "nextunique" / "prevunique", or both?

**Comment by yathit** [comment]
@inexorabletash both cases.

**Comment by inexorabletash** [comment]
sgtm!

**Comment by inexorabletash** [comment]
In the `openCursor()` case, is the lower-bound of the key range implicitly the index key if primaryKey is specified (and thus it's an error if a open lower bound is specified), or should the range and starting position be decoupled and require both the key and primaryKey to be explicit specified e.g. `openCursor(null, null, {key: x, primaryKey: y})` ?

**Comment by yathit** [comment]
Agree. `IDBIndexOpenCursorOptions` should be a complete cursor position.

    dictionary IDBIndexOpenCursorOptions {
      IDBKey key, // nullable, default to lower-bound
      IDBKey primaryKey; // nullable, default to first
    }

Using `openCursor` becomes easier. Let the keyRange unchanged and open with different starting position, without requiring keyRange to be met with cursor position.

**Comment by inexorabletash** [comment]
Should we throw if `primaryKey` is specified but `key` is not? (I vote yes.)

Should we define any special behavior if options are used on a non-multiEntry index? (I vote no.)

**Comment by inexorabletash** [comment]
For posterity, here's how you'd write a fallback if these methods didn't exist:

```js
  var options = {key: 2, primaryKey: 2};

  var req = idx.openCursor(null, 'next', options);
  req.onsuccess = function() {
    var cursor = req.result;
    if (!cursor) return;

    if (indexedDB.cmp(cursor.key, options.key) < 0 ||
      (indexedDB.cmp(cursor.key, options.key) === 0 &&
        indexedDB.cmp(cursor.primaryKey, options.primaryKey) < 0)) {

          // if you have continuePrimaryKey but not openCursor(..., options) for some reason
          if (cursor.continuePrimaryKey) {
            cursor.continuePrimaryKey(options.key, options.primaryKey);
            return;
          }

          // If index key is too low we can jump directly to that.
          if (indexedDB.cmp(cursor.key, options.key) < 0) {
            cursor.continue();
            return;
          }

          // Otherwise we're stuck iterating until we hit the right primary key.
          cursor.continue();
          return;
    }

    // The cursor is positioned at appropriate record - use it!
    cursor.continue();
  };
```

**Comment by pwnall** [comment]
Should this issue be closed? I'm pretty sure Chrome, Firefox and Safari are shipping this.

**Comment by inexorabletash** [comment]
`openCursor()` with _primaryKey_ specified is still a proposal - no-one implements that.

Currently you need to open the cursor against _key_ then maybe call `continuePrimaryKey()` to advance it to the right _primaryKey_ on the first iteration.

**Comment by pwnall** [comment]
Thanks for explaining! I missed that first part.
